File: backend/app/core/roles.py
"""Single source of truth for role and permission names.

Mirrored (identical string values) in `frontend/src/lib/roles.ts`. When you
rename or add a role/permission here, update that file too, plus the
native postgres `user_role` enum type in the database (see
`backend/sql/001_role_enum.sql` for how to add/rename a label -- it cannot
just be widened like a varchar). Nothing else in the backend should
reference these strings literally.
"""

from enum import StrEnum


class Role(StrEnum):
    ORGANISER = "organiser"  # Event Organiser -- highest privilege, admin-equivalent
    COORDINATOR = "coordinator"  # Event Coordinator
    VENUE_STAFF = "venue_staff"  # Venue Staff
    TECH_SUPPORT = "tech_support"  # Technical Support Staff (12 chars -- exactly at the limit)
    ATTENDEE = "attendee"  # Attendee


# Human-readable label for each role, since the slugs above are not
# presentable as-is (e.g. "venue_staff" -> "Venue Staff"). Used by the UI;
# never compared against or stored.
ROLE_LABELS: dict[Role, str] = {
    Role.ORGANISER: "Event Organiser",
    Role.COORDINATOR: "Event Coordinator",
    Role.VENUE_STAFF: "Venue Staff",
    Role.TECH_SUPPORT: "Technical Support Staff",
    Role.ATTENDEE: "Attendee",
}


class Permission(StrEnum):
    # One member per capability, grouped by domain.
    EVENT_READ = "event:read"
    EVENT_WRITE = "event:write"
    EVENT_APPROVE = "event:approve"
    VENUE_READ = "venue:read"
    VENUE_BOOK = "venue:book"
    VENUE_MANAGE = "venue:manage"
    EQUIPMENT_READ = "equipment:read"
    EQUIPMENT_MANAGE = "equipment:manage"
    REGISTRATION_READ = "registration:read"
    REGISTRATION_MANAGE = "registration:manage"


# Explicit per-role permission table. ORGANISER gets every permission (it
# is the admin-equivalent -- there is no separate admin role); each lower
# tier is written out explicitly rather than inheriting implicitly from
# the tier above it, so the whole grant surface for a role is visible in
# one place and easy to audit.
ROLE_PERMISSIONS: dict[Role, frozenset[Permission]] = {
    Role.ORGANISER: frozenset(Permission),
    Role.COORDINATOR: frozenset(
        {
            Permission.EVENT_READ,
            Permission.EVENT_WRITE,
            Permission.VENUE_READ,
            Permission.VENUE_BOOK,
            Permission.EQUIPMENT_READ,
            Permission.REGISTRATION_READ,
            Permission.REGISTRATION_MANAGE,
        }
    ),
    Role.VENUE_STAFF: frozenset(
        {
            Permission.VENUE_READ,
            Permission.VENUE_MANAGE,
            Permission.EVENT_READ,
        }
    ),
    Role.TECH_SUPPORT: frozenset(
        {
            Permission.EQUIPMENT_READ,
            Permission.EQUIPMENT_MANAGE,
            Permission.EVENT_READ,
        }
    ),
    Role.ATTENDEE: frozenset(
        {
            Permission.EVENT_READ,
            Permission.VENUE_READ,
            Permission.REGISTRATION_READ,
        }
    ),
}

# Public registration must only ever create attendees; every other role is
# granted by an ORGANISER via PATCH /users/{id}/role.
DEFAULT_ROLE = Role.ATTENDEE


def role_has(role: Role, permission: Permission) -> bool:
    """Return whether `role` grants `permission`."""
    return permission in ROLE_PERMISSIONS.get(role, frozenset())

File: frontend/src/lib/roles.ts
// Mirror of backend/app/core/roles.py — keep string values byte-identical.
// This map is for *rendering* only; the backend is the real gate on every
// request. When a role/permission is renamed, edit both files (plus the
// SQL CHECK constraint / UPDATE for any stored `role` values, see
// backend/sql/001_role_constraint.sql).

export const Role = {
  ORGANISER: 'organiser',
  COORDINATOR: 'coordinator',
  VENUE_STAFF: 'venue_staff',
  TECH_SUPPORT: 'tech_support',
  ATTENDEE: 'attendee',
} as const
export type Role = (typeof Role)[keyof typeof Role]

// Human-readable label for each role — the slugs above are not
// presentable as-is (e.g. "venue_staff" -> "Venue Staff").
export const ROLE_LABELS: Record<Role, string> = {
  [Role.ORGANISER]: 'Event Organiser',
  [Role.COORDINATOR]: 'Event Coordinator',
  [Role.VENUE_STAFF]: 'Venue Staff',
  [Role.TECH_SUPPORT]: 'Technical Support Staff',
  [Role.ATTENDEE]: 'Attendee',
}

export const Permission = {
  EVENT_READ: 'event:read',
  EVENT_WRITE: 'event:write',
  EVENT_APPROVE: 'event:approve',
  VENUE_READ: 'venue:read',
  VENUE_BOOK: 'venue:book',
  VENUE_MANAGE: 'venue:manage',
  EQUIPMENT_READ: 'equipment:read',
  EQUIPMENT_MANAGE: 'equipment:manage',
  REGISTRATION_READ: 'registration:read',
  REGISTRATION_MANAGE: 'registration:manage',
  USER_READ: 'user:read',
  ROLE_ASSIGN: 'role:assign',
} as const
export type Permission = (typeof Permission)[keyof typeof Permission]

export const ROLE_PERMISSIONS: Record<Role, readonly Permission[]> = {
  [Role.ORGANISER]: Object.values(Permission),
  [Role.COORDINATOR]: [
    Permission.EVENT_READ,
    Permission.EVENT_WRITE,
    Permission.VENUE_READ,
    Permission.VENUE_BOOK,
    Permission.EQUIPMENT_READ,
    Permission.REGISTRATION_READ,
    Permission.REGISTRATION_MANAGE,
  ],
  [Role.VENUE_STAFF]: [Permission.VENUE_READ, Permission.VENUE_MANAGE, Permission.EVENT_READ],
  [Role.TECH_SUPPORT]: [Permission.EQUIPMENT_READ, Permission.EQUIPMENT_MANAGE, Permission.EVENT_READ],
  [Role.ATTENDEE]: [Permission.EVENT_READ, Permission.VENUE_READ, Permission.REGISTRATION_READ],
}

export function roleHas(role: Role, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role]?.includes(permission) ?? false
}

// Single source of truth for "which page does this role land on" — used by
// the /my redirect route and by Dashboard's link to the user's own page.
export const ROLE_HOME_PATH: Record<Role, string> = {
  [Role.ORGANISER]: '/organiser',
  [Role.COORDINATOR]: '/coordinator',
  [Role.VENUE_STAFF]: '/venue-staff',
  [Role.TECH_SUPPORT]: '/tech-support',
  [Role.ATTENDEE]: '/attendee',
}
